### PR TITLE
Plumb ValuesSourceRegistry through to QuerySearchContext

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR gets rid of the singleton implementation of the `ValuesSourceRegistry` I had been using for prototyping, and replaces it with something that behaves like we generally expect a registry to behave.  `SearchModule` creates an instance of `ValuesSourceRegistry`, which then gets pulled down through the various index management classes, until it gets made available to the aggregations framework via `QueryShardContext`.

This also lets us get rid of the atomic initialization guards on the aggregation builders, since tests creating multiple `SearchModule` instances now get isolated `ValuesSourceRegistry` instances.

Special thanks to @nik9000 for pointing me in the right direction on this!

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexModule.java [hunk 1]
```diff
@@ -382,21 +383,21 @@ public static Type defaultStoreType(final boolean allowMmap) {
...
+        BigArrays bigArrays,
+        ThreadPool threadPool,
+        ScriptService scriptService,
+        ClusterService clusterService,
+        Client client,
+        IndicesQueryCache indicesQueryCache,
+        MapperRegistry mapperRegistry,
+        IndicesFieldDataCache indicesFieldDataCache,
+        NamedWriteableRegistry namedWriteableRegistry,
+        BooleanSupplier idFieldDataEnabled, ValuesSourceRegistry valuesSourceRegistry)
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/IndexModule.java]: I think the old indentation was easier to read.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 2]
```diff
@@ -239,12 +241,13 @@ public IndicesService(Settings settings, PluginsService pluginsService, NodeEnvi
                           IndexScopedSettings indexScopedSettings, CircuitBreakerService circuitBreakerService, BigArrays bigArrays,
                           ScriptService scriptService, ClusterService clusterService, Client client, MetaStateService metaStateService,
                           Collection<Function<IndexSettings, Optional<EngineFactory>>> engineFactoryProviders,
-                          Map<String, IndexStorePlugin.DirectoryFactory> directoryFactories) {
+                          Map<String, IndexStorePlugin.DirectoryFactory> directoryFactories, ValuesSourceRegistry valuesSourceRegistry) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java]: It look like someone was going with one arg per line here.

File: server/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 3]
```diff
@@ -298,6 +299,8 @@
      */
     public SearchModule(Settings settings, List<SearchPlugin> plugins) {
         this.settings = settings;
+        //TODO: constructor call goes here
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java]: Leftover TODO?

File: server/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 4]
```diff
@@ -298,6 +299,8 @@
...
+        this.valuesSourceRegistry = new ValuesSourceRegistry();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java]: Do you still like the idea of changing the ctor to take an `Iterable`? Is that a thing for a follow up change?

**Review by nik9000 — APPROVED** [review]
I left a question inline and a couple of formatting requests for if you feel like making me happy. Otherwise LGTM. It certainly feels good to remove the static singleton!

File: server/src/main/java/org/elasticsearch/index/IndexModule.java (hunk 1 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/index/IndexModule.java]: Yeah, not sure what happened here.  This is neither the old indentation, nor the format we're standardizing on.  I'll fix it.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 2 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java]: I'm a little confused by this comment - are you saying this should be one line per arg? I think that's what the new auto-formatting would do.

File: server/src/main/java/org/elasticsearch/search/SearchModule.java (hunk 4 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java]: I didn't see a clean way to do this.  We already manually list all the aggregations once in `registerAggregations`, and that still feels like the right place to hook in the values source mappings for each aggregation.  To do that in the constructor, I'd basically need to duplicate that list.  I thought briefly about some kind of visitor pattern to cover both uses, but it seemed like a lot of boiler plate that didn't add any clarity.  If you've got a good idea for how we could make this cleaner, I'm happy to work with you on it, but I don't see a better way than what we're doing now.

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/indices/IndicesService.java]: When I first looked it at I thought that *was* what it was and you hadn't followed. On second look it wasn't. Sorry! Ignore me.

File: server/src/main/java/org/elasticsearch/search/SearchModule.java (hunk 4 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java]: I think it is cool to merge as is and grab that in a follow up. Would it be ok if I gave it a shot myself once this is in?

Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java]: Absolutely!

**Comment by not-napoleon** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by not-napoleon** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Review by polyfractal — APPROVED** [review]
Late to the party, but yay this feels so much nicer :tada: :)

**Pull request merged**